### ISSUE TITLE
Snippet Campaign Block Logic Causing Auto-Block, so user never sees snippet

### ISSUE THREAD

**Issue by stsegovia-moz** [mozmeao/snippets-service#911]
**TL;DR:** Snippets aren't being show to all users, only some. We identified traffic via Google Analytics, but can't see the snippets on several computers. We investigated and found that snippets were being blocked via campaigns, even when users had not previously interacted with the snippet.

**Long Version**

We are having issues with Snippets not displaying for all users. This is an ongoing issue that we've identified. Most recently on Feb 4th, we published a snippet for IRL podcast Season 4 Episode 5. 

**On 3 different computers _with snippets enabled_, we could not see any of the 3 snippets related to the IRL podcast campaign.**

Upon deeper investigation in Google Analytics, we confirmed that the snippets were delivering traffic to the IRL podcast landing page. Thus, we could deduce that some users were able to see and interact with these snippets -- yet we still couldn't see the snippets.

Then, we checked the _about:newtab#asrouter_ page to see if the IRL snippets were blocked on one of the computers. In fact, the snippets were already blocked. [see attached screenshot]

![screen shot 2019-02-06 at 10 45 51 am](https://user-images.githubusercontent.com/47153345/52383589-e035df80-2a2e-11e9-9e6d-de3f586b8f82.png)

As a test, we removed the campaign from each snippet, re-publishing the snippets and leaving the _campaign_ field blank. After waiting a few hours, we were able to see the snippets on on the same computer where we originally observed the snippets being blocked. This seems like a temporary solution for this issue. 

I'm deducing that certain snippet campaigns are being blocked before a user has a chance to interact with any snippet within that campaign. They can only be unblocked if we 1) Remove campaign name, or 2) Manually unblock snippets via the _about:newtab#asrouter_ page. 

Can you please provide guidance? Thank you!

**Below are the snippet details.**

- **Who:** All Snippet Users with snippets enabled
- **What:** Campaigns being blocked before users even interact with them
- **Snippet IDs:** 10580, 10581, 10582
- **Campaign:** IRL_S4_E5
- 
- **-Steve**

**Comment by glogiotatidis** [comment]
Hi Steve and thanks for the very detailed report. 

The cases were a snippets gets blocked are the following:

 - User clicks on X button
 - User interacts with the snippet, e.g. visits link, subscribes to newsletter etc

In both cases above, when `campaign` is defined all the snippets with the same campaign slug are also blocked. This is true for future snippets using the same campaign 

So for example:
 - Snippet A with campaign name `foo` gets published
 - User clicks the link on the snippet, Snippet A gets blocked
 - Snippet B with campaign name `foo` gets published
 - User will never get see Snippet B. It will appear in the `about:newtab#asrouter` page as blocked.

I'll continue to investigate a lead and report here

**Comment by glogiotatidis** [comment]
I think I've figured this out: It seems that the blocking functionality gets triggered in AS even when in preview mode. This means that when you interact with the previewed snippet or click the X button then the snippet and all the snippets with the same campaign get blocked . At this state the snippet does not appear in the `about:newtab#asrouter` page, because the page does not list preview snippets.

In your case when the snippet got published, it was already in your blocklist. So you never got the snippet and when you inspected `about:newtab#asrouter` page it was there, marked blocked.

Removing (or changing) the campaign name works b/c the blocking mechanism uses the campaign name when available and if not defaults to snippet ID.

Do the above sound a reasonable explanation to what you're expieriecing @stsegovia-moz ?


While this is an AS bug we can workaround this by forcing empty campaigns and different ID numbers when previewing the snippet.

**Comment by stsegovia-moz** [comment]
Hi Giorgos,

Yes, that does seem like a reasonable explanation for Jean and I, since we tend to preview snippets often. 

But, we had a fellow Mozillian from a different team, who does not interact with the snippets preview, that experience this issue as well. She was actually the one who reached out to our team because she couldn't see these snippets that were part of her project. We haven't asked to check her blocked snippets in `about:newtab#asrouter` page yet, but we can investigate and confirm that via a computer that has not blocked any snippets to see if they experience 1) the issue of not seeing snippets and 2) that they are blocked on the `about:newtab#asrouter` page

Jean had the same thought as you did regarding an issue differentiation between blocking the snippet preview and blocking published snippets, but from feedback from our peers who have not interacted with snippets (or previews) prior, it sounds like it's possible it may be another cause.  

We'll see if we can investigate further on our part too. Thanks Giorgos!

**Comment by glogiotatidis** [comment]
> But, we had a fellow Mozillian from a different team, who does not interact with the snippets preview, that experience this issue as well.

Can we confirm that the user hasn't been given the preview link at some point before going live?

**Comment by stsegovia-moz** [comment]
Hi Giorgos,

Yes, I can confirm that this individual has not seen a preview link prior to the snippet going live (or at any point). I can't speculate what may cause that, but we have had similar reports recently from individuals who do not interact with snippet preview links.

**Comment by glogiotatidis** [comment]
I will go ahead and fix the preview issue blocking and will comment here when that's deployed in production. 

After that if we experience this issue again we'll need as much info as possible to be able to debug further.

**Comment by stsegovia-moz** [comment]
Sounds great! Thank you Giorgos!

**Comment by glogiotatidis** [comment]
The fix is now in production. 

If users experience this behavior again with snippets created after today, please re-open this issue to further explore.

Thanks!
